### DISCUSSION TITLE
Go client API?

### DISCUSSION THREAD

**Discussion by paulbowler** [pocketbase/pocketbase#5700] [Q&A]
I'm looking at creating my app in Go + HTMX using additional routes configured directly within the pocketbase main.go

This will would allow me to combine both client and server into a single codebase.

Something like this:

```
func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {

		e.Router.GET("/app/*", func(c echo.Context) error {
			tmpl := template.Must(template.ParseFiles("templates/index.html"))
			return tmpl.Execute(c.Response().Writer, nil)
		})
		
		... (more routes)
	})

	// Start the Pocketbase app
	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

However, this appears to mean that I no longer have access to the client API as I would using Javascript, only much lower level framework and database functions that require me to code authentication and authorisation functions manually.

Am I missing something? Is there a way to use the API from within the same app codebase?

Thanks,

Paul

**Comment by Plorenzo** [comment]
You should read the docs

https://pocketbase.io/docs/go-records/

**Reply by paulbowler** [reply]
Indeed, I'm getting very familiar with the documentation.

I'm really looking for a way to manage authentication and authorisation with as little coding as possible. In the JS API it can be done easily using:

```
const authData = await pb.collection('users').authWithPassword('YOUR_USERNAME_OR_EMAIL', '1234567890');

// after the above you can also access the auth data from the authStore
console.log(pb.authStore.isValid);
console.log(pb.authStore.token);
console.log(pb.authStore.model.id);

// "logout" the last authenticated model
pb.authStore.clear();
```

Is there any equivalent functions I can use directly in server-side go?

P.

**Comment by thangngoc89** [comment]
There isn't anything official but for 3rd party, there was some work on Go client last year. You could check the discussion thread at https://github.com/pocketbase/pocketbase/discussions/995 
 and the repo at https://github.com/r--w/pocketbase

**Comment by whmeitzler** [comment]
You can include the Pocketbase client lib as static files or custom routes.  

The tricky part for HTMX seems to be auth -- how do you know who an HTMX request came from? The best answer I have found for this is to hook into HTMX's calls, passing along the auth token from the Pocketbase client lib as seen [here](https://github.com/ccutch/pbj/blob/9738b54e475e2bd29612fd9fe308552b330a92dd/pages.go#L109https://github.com/ccutch/pbj/blob/9738b54e475e2bd29612fd9fe308552b330a92dd/pages.go#L109) (Thanks [ccutch](https://github.com/ccutch))

so the base page would include something like:
```html
 <script>
	  const pb = new PocketBase(window.location.href);
      document.body.addEventListener("htmx:configRequest", function (event) {
	    if (pb && pb.authStore) event.detail.headers['Authorization'] = "Bearer "+pb.authStore.token;
      });
    </script>
```
then all `hx-post` events on the page would carry the bearer token that Pocketbase expects. 

You'd still need to figure out how to get the username and password into the `pb` instance. 

It would be interesting to figure out a proper HTMX middleware to solve auth.

**Reply by pnmcosta** [reply]
This is the approach, I've used it before with oauth and worked flawlessly. Regarding:

> You'd still need to figure out how to get the username and password into the pb instance.
> It would be interesting to figure out a proper HTMX middleware to solve auth.

I don't think you would need to, `pb.authStore` is localStorage which persists on browser recycle, so `const authData = await pb.collection('users').authWithPassword('YOUR_USERNAME_OR_EMAIL', '1234567890');` should be the way as per OP, I believe.

Granted this is putting auth state in the client, and manage it there instead, I would probably recurrently have a token check for reauth.
